tests/unit/cluster/slot-migration.tcl

```diff
@@ -0,0 +1,363 @@
+        wait_for_slot_state 3 "\[7->-$R1_id\] \[13->-$R1_id\] \[17->-$R1_id\]"
+        wait_for_slot_state 4 "\[7-<-$R0_id\] \[13-<-$R0_id\] \[17-<-$R0_id\]"
+    }
+
+    test "New replica inherits multiple migrating slots" {
+        # Reset R3 to turn it into an empty node
+        assert_equal {OK} [R 3 CLUSTER RESET]
+        # Add R3 back as a replica of R0
+        assert_equal {OK} [R 3 CLUSTER MEET [srv 0 "host"] [srv 0 "port"]]
+        wait_for_role 0 master
+        assert_equal {OK} [R 3 CLUSTER REPLICATE $R0_id]
+        wait_for_role 3 slave
+        # Validate final states
+        wait_for_slot_state 3 "\[7->-$R1_id\] \[13->-$R1_id\] \[17->-$R1_id\]"
+    }
+
+    test "Slot finalization succeeds on both primary and replicas" {
+        assert_equal {OK} [R 1 CLUSTER SETSLOT 7 NODE $R1_id]
+        wait_for_slot_state 1 "\[13-<-$R0_id\] \[17-<-$R0_id\]"
+        wait_for_slot_state 4 "\[13-<-$R0_id\] \[17-<-$R0_id\]"
+        assert_equal {OK} [R 1 CLUSTER SETSLOT 13 NODE $R1_id]
+        wait_for_slot_state 1 "\[17-<-$R0_id\]"
+        wait_for_slot_state 4 "\[17-<-$R0_id\]"
+        assert_equal {OK} [R 1 CLUSTER SETSLOT 17 NODE $R1_id]
+        wait_for_slot_state 1 ""
+        wait_for_slot_state 4 ""
+    }
+
+}
+
+start_cluster 3 3 {tags {external:skip cluster} overrides {cluster-allow-replica-migration no cluster-node-timeout 1000} } {
+
+    set node_timeout [lindex [R 0 CONFIG GET cluster-node-timeout] 1]
+    set R0_id [R 0 CLUSTER MYID]
+    set R1_id [R 1 CLUSTER MYID]
+
+    test "Slot is auto-claimed by target after source relinquishes ownership" {
+        migrate_slot 0 1 609
+        #Validate that R1 doesn't own slot 609
+        catch {[R 1 get aga]} e
+        assert_equal {MOVED} [lindex [split $e] 0]
+        #Finalize the slot on the source first
+        assert_equal {OK} [R 0 CLUSTER SETSLOT 609 NODE $R1_id]
+        after $node_timeout
+        #R1 should claim slot 609 since it is still importing slot 609
+        #from R0 but R0 no longer owns this slot
+        assert_equal {OK} [R 1 set aga foo]
+    }
+}
+
+start_cluster 3 3 {tags {external:skip cluster} overrides {cluster-allow-replica-migration no cluster-node-timeout 1000} } {
+    set R1_id [R 1 CLUSTER MYID]
+
+    test "CLUSTER SETSLOT with an explicit timeout" {
+        # Simulate a replica crash
+        catch {R 3 DEBUG SEGFAULT} e
+
+        # Setslot with an explicit 1ms timeoout
+        set startTime [clock milliseconds]
+        catch {R 0 CLUSTER SETSLOT 609 MIGRATING $R1_id TIMEOUT 3000} e
+        set endTime [clock milliseconds]
+        set duration [expr {$endTime - $startTime}]
+
+        # Assert that the execution time is greater than the default 2s timeout
+        assert {$duration > 2000}
+
+        # Setslot should fail with not enough good replicas to write after the timeout
+        assert_equal {NOREPLICAS Not enough good replicas to write.} $e
```

Comment:
So, in this case the slot on one of the replica will be marked as importing right? Does that ever get cleared?

Reply:
>in this case the slot on one of the replica will be marked as importing right? 

correct

>Does that ever get cleared?

No. I am not implementing 2PC here. The slot migration protocol itself will never be self-sufficient. The user needs to make sure to set `importing` on the target first and only upon successful return set `migrating` on the source. This way, we can guarantee that a correct forwarding link is maintained all the time and this becomes only a performance issue (due to unnecessary forwarding) but not a correctness issue. More specifically

1. we have partial `importing` states on the target shard

 no impact because without a matching `migrating` state on the source, the forwarding won't occur
 
 2. we have partial `migrating` states on the source shard
 
this implies that a failure would be returned to the caller for its `cluster setslot` call. if a failover occurs afterwards, we would establish a forwarding link unexpectedly (to the caller). however, I think this is still fine because there would be no keys in the target shard hence a `get` request would fail in the same way (except the additional network hop). the `set` request though would be, again unexpectedly, forwarded to the target and the new key end up on the target shard. This is not desired but should be acceptable. it is mostly correct behavior with just one violation of the contract (lying about the `cluster setslot` failure`). 

also in the future when we introduce atomic slot migration, I would like the engine to automatically resume the incomplete slot migration and we will not end up in this intermediate state forever. we don't have to make a decision on the atomic slot migration behavior here but just wanted to share some additional thoughts.

Reply:
Yeah, it's more weird behavior, but for now it seems acceptable.

Reply:
We'll have to accept it until atomic slot migration, I suppose.

Reply:
Wait a second. Now the command fails on timeout? From earlier discussions, I though this was best effort.

What will `valkey-cli` do if SETSLOT MIGRATING or IMPORING fails?

What will `valkey-cli` do if SETSLOT NODE fails?

Is the slot migration finished or not?